### ISSUE TITLE
bug(android/engine): StringIndexOutOfBoundsException when end > length

### ISSUE THREAD

**Issue by darcywong00** [keymanapp/keyman#11160]
Relates to #11128 which fixed StringIndexOutOfBoundsException when the selection was reversed

@mcdurdin [triggered this similar crash](https://keyman.sentry.io/issues/5022225846/events/6bd1618d50204522bb5f8c0fe15a5a43) while repeatedly holding down backspace. It's the same area of code, but where the substring end is > length

```javascript
StringIndexOutOfBoundsException
begin 0, end 12, length 11

java.lang.String in checkBoundsBeginEnd at line 4466
java.lang.String in substring at line 2517
com.keyman.engine.KMKeyboard in updateSelectionRange at line 190
com.keyman.engine.KMManager in updateSelectionRange at line 2150
com.keyman.android.SystemKeyboard in onUpdateSelection at line 132
android.inputmethodservice.InputMethodService$InputMethodSessionImpl in updateSelection at line 1414
android.inputmethodservice.IInputMethodSessionWrapper in executeMessage at line 111
com.android.internal.os.HandlerCaller$MyHandler in handleMessage at line 44
android.os.Handler in dispatchMessage at line 106
android.os.Looper in loopOnce at line 230
android.os.Looper in loop at line 319
android.app.ActivityThread in main at line 8893
java.lang.reflect.Method in invoke
com.android.internal.os.RuntimeInit$MethodAndArgsCaller in run at line 608
com.android.internal.os.ZygoteInit in main at line 1103
```

I think we should catch this, but also add a limit
end = (end > length) ? length : end;

And if possible add some unit test

**Comment by jahorton** [comment]
Investigation of the issue yields something potentially enlightening:  this first started showing up with 17.0.280-beta.  The thing most likely to have affected this: #10885, [which landed then](https://github.com/keymanapp/keyman/blob/master/HISTORY.md#170280-beta-2024-03-01).

Someone (could be me, could be @darcywong00) would need to dig into where the underlying assumptions are invalid in order to get a solid fix, I think.

**Comment by darcywong00** [comment]
Happy for you to spelunk on this one. afict

The consistent crash is the `substring()` in l.190 where selStart > rawText.length()

https://github.com/keymanapp/keyman/blob/93dc41e2f5e1892f0de3fa821f44da8fbd144355/android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java#L166-L192

Is `selStart` coming from KeymanWeb? The behavior is that the Android app (rawText) is deleting faster than KeymanWeb is updated.

**Comment by darcywong00** [comment]
Handing off to @jahorton for spelunking

**Comment by jahorton** [comment]
Well, I found the invalid assumption.  We have ourselves a concurrency issue.  It is possible for the object returned by `.getExtractedText` to have a more recent update to selection range than the call that triggered our check against `.getExtractedText`.  Fortunately, the returned `ExtractedText` object itself provides `.selectionStart` and `.selectionEnd` fields, and these will be current.  (These are also the proof of asynchronicity as seen in the following screenshots.)

`KMKeyboard.updateSelectionRange` and its parameters during a rapid backspacing sequence, synchronous with our handler for `onUpdateSelection`:

![selStart == selEnd == 711](https://github.com/keymanapp/keyman/assets/25213402/f5310a5e-0c0b-49d1-891a-e9e072ed42fd)

Same call, but checking the values available via the `ExtractedText` object:

![extractedSelStart == extractedSelEnd == 710](https://github.com/keymanapp/keyman/assets/25213402/06f7420e-ff55-498b-87a3-49b0d5b3703d)

As the `InputMethodService` interface with the editor and our extension is asynchronously managed, updates can happen during handling of raised events.  The call still serves well as a signal that "something has updated", but we need an atomic operation to retrieve all relevant values simultaneously in order to avoid an issue here - and `getExtractedText` looks to be sufficient in this regard.

**Comment by darcywong00** [comment]
Fixed by #11188
